src/exporter/exporter/csv.go

```diff
@@ -25,131 +20,123 @@ var fieldUnits = map[blackbox.FieldName]string{
-		e.NumberOfFramesRead++
-		var values []string
-		for k, v := range frame.Values().([]int32) {
-			if i, _ := e.frameDef.GetFieldIndex(blackbox.FieldVbatLatest); k == i {
-				vbat := (float64(v) * adcVref * 10.0 * float64(e.frameDef.Sysconfig.Vbatscale)) / 4095.0
-				values = append(values, prependSpaceForField(k, fmt.Sprintf("%.3f", math.Floor(vbat)/1000.0)))
-				continue
-			}
-			if i, _ := e.frameDef.GetFieldIndex(blackbox.FieldAmperageLatest); k == i {
-				millivolts := float64((uint32(v)*adcVref*100)/4095) - float64(e.frameDef.Sysconfig.CurrentMeterOffset)
-				millivolts = (millivolts * 10000) / float64(e.frameDef.Sysconfig.CurrentMeterScale)
-				e.state.currentMilliamps = millivolts
-				values = append(values, prependSpaceForField(k, fmt.Sprintf("%.3f", millivolts/1000)))
-				continue
-			}
-			values = append(values, prependSpaceForField(k, fmt.Sprintf("%d", v)))
-		}
+		values := e.friendlyMainFrameValues(frame.(*blackbox.MainFrame).Values().([]int32))
 
-		if e.state.hasAmperageAdc {
-			if e.state.lastTime != 0.0 {
-				e.state.energyMilliampHours += (e.state.currentMilliamps * float64(frame.Values().([]int32)[1]-e.state.lastTime)) / (time.Hour.Seconds() * 1000000)
-			}
-			e.state.lastTime = frame.Values().([]int32)[1]
-
-			values = append(values, prependSpaceForField(0, fmt.Sprintf("%f", e.state.energyMilliampHours)))
+		err := e.writeLn(strings.Join(values, ", "))
+		if err != nil {
+			return errors.Wrapf(err, "could not write frame '%s' to target file", string(frame.Type()))
 		}
+	}
+	return nil
+}
 
-		for _, v := range e.lastSlow.StringValues() {
-			values = append(values, v)
+func (e *CsvFrameExporter) friendlyMainFrameValues(valuesS []int32) []string {
+	var values []string
+	for k, v := range valuesS {
+		if i, _ := e.frameDef.GetFieldIndex(blackbox.FieldVbatLatest); k == i {
+			e.batteryState.setLatestVbat(v)
+			values = append(values, prependSpaceForField(k, fmt.Sprintf("%.3f", math.Floor(e.batteryState.voltageVolt*1000)/1000)))
+			continue
 		}
-		_, err := e.target.Write([]byte(strings.Join(values, ", ")))
-		if err != nil {
-			return errors.Wrapf(err, "could not write frame '%s' to target file", string(frame.Type()))
+		if i, _ := e.frameDef.GetFieldIndex(blackbox.FieldAmperageLatest); k == i {
+			timeFieldIndex, _ := e.frameDef.GetFieldIndex(blackbox.FieldTime)
+			e.batteryState.setLatestAmperage(v, valuesS[timeFieldIndex])
+			values = append(values, prependSpaceForField(k, fmt.Sprintf("%.3f", e.batteryState.currentAmps)))
+			continue
 		}
+		values = append(values, prependSpaceForField(k, fmt.Sprintf("%d", v)))
+	}
+
+	if e.hasAmperageAdc {
+		values = append(values, prependSpaceForField(0, fmt.Sprintf("%f", e.batteryState.energyMilliampHours)))
 	}
-	return e.writeNewLine()
+
+	for _, v := range e.lastSlowFrame.StringValues() {
+		values = append(values, v)
+	}
+	return values
+}
+
+func (e *CsvFrameExporter) writeBytes(data []byte) error {
+	_, err := e.target.Write(data)
+	return err
```

Comment:
I still suggest to wrap it with `errors.Wrap` or `errors.WithStack`, to have a call-stack